### PR TITLE
docs: Spring Boot Tutorial v1

### PR DESCRIPTION
**Pull request by nnennandukwe**
## About the changes

We have a v1 for a Java Spring Boot Tutorial:
- uses popular Java Spring Boot open source project Spring Pet Clinic
- uses Unleash Spring Boot SDK
- this is simple & introductory as we don't have data plugged into the new page we build in the application.

spring-boot-tutorial-app-in-browser

Closes #

## Discussion points

I added the tutorial to the Java section in the docs navigation menu. I could have it stand alone, but I would consider it to be underneath the Java language we have already listed in our menu.

Screenshot 2024-03-19 at 8 49 11 PM

### CODE REVIEW

File: website/docs/feature-flag-tutorials/java/spring-boot-implementing-feature-flags.md [hunk 1]
```diff
@@ -0,0 +1,384 @@
+---
+title: How to Implement Feature Flags in Java Spring Boot
+slug: /feature-flag-tutorials/spring-boot
+---
+
+Java [Spring Boot](https://spring.io/projects/spring-boot) is a popular Java framework for getting apps up and running with minimal configuration. It is often used to develop microservices for large enterprise software.
+
+Leveraging feature flags allows developers to toggle new features on and off, whether you’re experimenting in your local environment, testing for QA purposes, or rolling out changes to users in production. With Unleash, an open-source feature flag service, you can use our tooling to implement feature flags into your application and release new features faster, strategically, and safely. But how can you do this in Java Spring Boot?
+
```
Comment (alvinometric) [inline website/docs/feature-flag-tutorials/java/spring-boot-implementing-feature-flags.md:9]: I would remove the first 2 paragraphs

File: website/docs/feature-flag-tutorials/java/spring-boot-implementing-feature-flags.md [hunk 2]
```diff
@@ -0,0 +1,384 @@
...
+## Prerequisites
+
+
+In this tutorial, you will need the following:
+- A web browser like Chrome or Firefox
+- Git
+- Docker
+- Maven or Gradle for building Java
+- (Optional) a code editor like IntelliJ IDEA
```
Comment (alvinometric) [inline website/docs/feature-flag-tutorials/java/spring-boot-implementing-feature-flags.md:29]: Capitalise the A ? not 100% sure about this one

File: website/docs/feature-flag-tutorials/java/spring-boot-implementing-feature-flags.md [hunk 3]
```diff
@@ -0,0 +1,384 @@
...
+
+![Java Spring Boot Architectural Diagram](/img/spring-boot-tutorial-architectural-diagram.png)
+
+This architecture diagram breaks down how the Java Spring Boot app works with Unleash to use feature flags.
+
+The Unleash Server is a Feature Flag Control Service for managing and storing your feature flags. It enables the retrieval of flag data and, particularly when not utilizing a user interface, supports sending data to and from the service. The Unleash Server has a UI for creating and managing projects and feature flags. API commands are also [available](/reference/api/unleash) to perform the same actions from your CLI or server-side app.
```
Comment (alvinometric) [inline website/docs/feature-flag-tutorials/java/spring-boot-implementing-feature-flags.md:35]: I would remove that paragraph too

File: website/docs/feature-flag-tutorials/java/spring-boot-implementing-feature-flags.md [hunk 4]
```diff
@@ -0,0 +1,384 @@
...
+
+Since Spring Boot is a framework used for Java backend apps, there are special security considerations to plan around when implementing feature flags.
+
+Most importantly, you must:
+- Limit feature flag payloads for scalability, security, and efficiency
+- Improve architectural resiliency with graceful degradation
+
+As your application scales, performance and resiliency become more critical and costly if not addressed. A feature flagging system should not be why your app slows down or fails. That’s why we recommend you account for this by reducing the size of your feature flag payloads. For example, instead of making one large call to retrieve flag statuses for all users as part of your configuration, group your users by specific attributes as part of your targeting rules that would be most relevant to your application.
+
+Additionally, you can cache your feature flag configuration to help reduce network round trips and dependency on external services. You can rely on the cache if your Feature Flag Control Service is unavailable, mitigating potential application failure.
```
Comment (alvinometric) [inline website/docs/feature-flag-tutorials/java/spring-boot-implementing-feature-flags.md:49]: I would also remove the previous 2 paragraphs there

File: website/docs/feature-flag-tutorials/java/spring-boot-implementing-feature-flags.md [hunk 5]
```diff
@@ -0,0 +1,384 @@
...
+![Create a flag in Unleash called "productsPageFlag".](/img/spring-boot-tutorial-create-flag.png)
+
+Your new feature flag has been created and is ready to be used. Enable the flag for your development environment, which makes it accessible for use in the Spring Boot app.
+
+
+![Enable the flag in the development environment in Unleash.](/img/spring-boot-tutorial-enable-flag.png)
+
+Next, generate an API token to use in your app. This API token will eventually be pulled into a configuration object within your Spring Boot app to toggle features.
+
+Note: We require an API token as part of your flag configuration to ensure only applications with the correct authentication can access your feature flags in Unleash. API tokens are created by users with API management access, and this controls how and by whom they can be distributed to applications that need them.
```
Comment (alvinometric) [inline website/docs/feature-flag-tutorials/java/spring-boot-implementing-feature-flags.md:98]: We can use the `:::note` syntax here if you think it's necessary

File: website/docs/feature-flag-tutorials/java/spring-boot-implementing-feature-flags.md [hunk 6]
```diff
@@ -0,0 +1,384 @@
...
+./mvnw spring-boot:run
+```
+
+If you want to use [Gradle](https://gradle.org/) as opposed to Maven to build the project, run the following command:
+
+```
+./gradlew build
+```
+
+For further details on building and running Spring Pet Clinic locally, refer to the [README on GitHub](https://github.com/spring-projects/spring-petclinic?tab=readme-ov-file).
```
Comment (alvinometric) [inline website/docs/feature-flag-tutorials/java/spring-boot-implementing-feature-flags.md:151]:
```suggestion
For further details on building and running Spring Pet Clinic locally, refer to the project's [README on GitHub](https://github.com/spring-projects/spring-petclinic?tab=readme-ov-file).
```

File: website/docs/feature-flag-tutorials/java/spring-boot-implementing-feature-flags.md [hunk 7]
```diff
@@ -0,0 +1,384 @@
...
+- A method that invokes the pet product service
+
+Next, we will create the `products.html` page returned from the `/products` endpoint.
+
+In `resources/templates/`, create `products.html` and use this small code snippet:
+
+
+```html
+<!DOCTYPE html>
+
```
Comment (alvinometric) [inline website/docs/feature-flag-tutorials/java/spring-boot-implementing-feature-flags.md:322]:
```suggestion
```

File: website/docs/feature-flag-tutorials/java/spring-boot-implementing-feature-flags.md [hunk 8]
```diff
@@ -0,0 +1,384 @@
...
+<html xmlns:th="https://www.thymeleaf.org" th:replace="~{fragments/layout :: layout (~{::body},'products')}">
+
```
Comment (alvinometric) [inline website/docs/feature-flag-tutorials/java/spring-boot-implementing-feature-flags.md:324]:
```suggestion
```

File: website/docs/feature-flag-tutorials/java/spring-boot-implementing-feature-flags.md [hunk 9]
```diff
@@ -0,0 +1,384 @@
...
+<body>
+   <h2>Clinic Products</h2>
+   <div th:text="${petProductsPage}"></div>
+</body>
+
```
Comment (alvinometric) [inline website/docs/feature-flag-tutorials/java/spring-boot-implementing-feature-flags.md:329]:
```suggestion
```

File: website/docs/feature-flag-tutorials/java/spring-boot-implementing-feature-flags.md [hunk 10]
```diff
@@ -0,0 +1,384 @@
...
+```
+
+Now that we have our services, controller, and an endpoint that maps to a page, we can add the product page to the app's navigation menu.
+
+In `layout.html`, insert a new list item on line 62.
+
+```html
+<li
+th:replace="~{::menuItem ('/products', 'products', 'view clinic products', 'th-list', 'Clinic products')}">
+<span class="fa fa-th-list" aria-hidden="true"></span>
```
Comment (alvinometric) [inline website/docs/feature-flag-tutorials/java/spring-boot-implementing-feature-flags.md:340]:
```suggestion
  <span class="fa fa-th-list" aria-hidden="true"></span>
```

File: website/docs/feature-flag-tutorials/java/spring-boot-implementing-feature-flags.md [hunk 11]
```diff
@@ -0,0 +1,384 @@
...
+<span>Clinic products</span>
```
Comment (alvinometric) [inline website/docs/feature-flag-tutorials/java/spring-boot-implementing-feature-flags.md:341]:
```suggestion
  <span>Clinic products</span>
```

File: website/docs/feature-flag-tutorials/java/spring-boot-implementing-feature-flags.md (hunk 1 above)
Comment (nnennandukwe) [reply website/docs/feature-flag-tutorials/java/spring-boot-implementing-feature-flags.md:9]: We've been using similar text as an introduction to our tutorials thus far.
An explanation of the language / framework we're using and then the importance of feature flags.
We'll be hearing more about ways to optimize these tutorials soon, which could mean minimizing these intros in the near future.

File: website/docs/feature-flag-tutorials/java/spring-boot-implementing-feature-flags.md (hunk 4 above)
Comment (nnennandukwe) [reply website/docs/feature-flag-tutorials/java/spring-boot-implementing-feature-flags.md:49]: We've been using this text for our tutorials to provide more perspective beyond just using a feature flag. It is meant to give devs insight into using feature flags at scale, in real world more complex environments so they're architecting it using some of our principles from our best practices documentation.

**Pull request merged**